### PR TITLE
[SYCL][E2E] Add granularity test for `sycl_ext_oneapi_virtual_mem` extension

### PR DESCRIPTION
**Pull request by DYNIO-INTEL**
Based on the test plan https://github.com/intel/llvm/pull/15509, this PR adds an e2e test checking memory granularities returned by `get_mem_granularity`.

### CODE REVIEW

File: sycl/test-e2e/VirtualMem/memory_granularity.cpp [hunk 1]
```diff
@@ -0,0 +1,76 @@
...
+
+int main() {
+  int Failed = 0;
+  sycl::queue Q;
+  sycl::context Context = Q.get_context();
+  sycl::device Device = Q.get_device();
+
+  Failed += CheckReturnedGranularitySize(
+      "returned granularity size must be greater then 0 for device and "
+      "granularity_mode::recommended",
```
Comment (AlexeySachkov) [inline sycl/test-e2e/VirtualMem/memory_granularity.cpp:35]:
```suggestion
      "recommended granularity size must be greater then 0 for device",
```

I think that we can simplify those messages like this, it should be mor or less obvious that we are talking about the value that is returned from some SYCL API

File: sycl/test-e2e/VirtualMem/memory_granularity.cpp [hunk 2]
```diff
@@ -0,0 +1,76 @@
+// REQUIRES: aspect-ext_oneapi_virtual_mem
```
Comment (AlexeySachkov) [inline sycl/test-e2e/VirtualMem/memory_granularity.cpp:1]: It probably make sense to outline that requirement into `lit.local.cfg`. See `VirtualFunctions` and `ESIMD` folders for examples

File: sycl/test-e2e/VirtualMem/memory_granularity.cpp [hunk 3]
```diff
@@ -0,0 +1,76 @@
...
+      syclext::get_mem_granularity(Device, Context,
+                                   syclext::granularity_mode::recommended),
+      syclext::get_mem_granularity(Context,
+                                   syclext::granularity_mode::recommended)};
+
+  std::array<size_t, NumberOfElements> MinimumGranularities = {
+      syclext::get_mem_granularity(Device, Context,
+                                   syclext::granularity_mode::minimum),
+      syclext::get_mem_granularity(Context,
+                                   syclext::granularity_mode::minimum)};
```
Comment (AlexeySachkov) [inline sycl/test-e2e/VirtualMem/memory_granularity.cpp:61]: Overall it seems like the whole test could be simplified to something like:

```c++
auto DevRecommended = syclext::get_mem_granularity(Device, Context,
                                   syclext::granularity_mode::recommended);
auto ContextRecommended = syclext::get_mem_granularity(Context,
                                   syclext::granularity_mode::recommended);
auto DevMinimum = syclext::get_mem_granularity(Device, Context,
                                   syclext::granularity_mode::minimum);
auto ContextMinimum = syclext::get_mem_granularity(Context,
                                   syclext::granularity_mode::minimum);
                                   
asssert(DevRecommended > 0 && "recommended granularity for device should be greater than 0");
asssert(ContextRecommended > 0 && "message");
// similar asserts for minimum
assert(DevRecommended >= DevMinimum && "message");
// similar assert for context
```

File: sycl/test-e2e/VirtualMem/memory_granularity.cpp [hunk 4]
```diff
@@ -0,0 +1,76 @@
+// REQUIRES: aspect-ext_oneapi_virtual_mem
+
+// This test checks whether memory granularity returned is greater then 0 and
```
Comment (maarquitos14) [inline sycl/test-e2e/VirtualMem/memory_granularity.cpp:3]:
```suggestion
// This test checks whether memory granularity returned is greater than 0 and
```

**Review by maarquitos14 — COMMENTED** [review]
Maybe I'm misreading the testplan, but it seems to me it says this test should be unittest not e2e test. @AlexeySachkov can you confirm?

**Comment by AlexeySachkov** [comment]
> Maybe I'm misreading the testplan, but it seems to me it says this test should be unittest not e2e test. @AlexeySachkov can you confirm?

Sorry for the confusion here. I should probably use a different terminology in the test plan, but due to the lack of a better one we have what we have. Suggestions are welcome, BTW!

Back to the question: even though the test plan says "unit", not all tests under that category are expected to be unit-tests in our regular understanding (i.e. put into `sycl/test` or `sycl/unittests`).

The main thing I wanted to highlight by the word "unit" is that we are not trying to test a proper meaningful usage scenario, but instead we just check internal consistency and correctness of certain APIs in isolation. That is done to have a better granularity in tests, so that when something breaks, we know the exact API which got broken and we don't have to debug a whole mini-application to understand why results are incorrect at the end.

This particular test should still stay E2E - moving it to `sycl/unittest` will mean that the only thing we would test are return values from `mock` plugin and not the actual values returned to us from low-level backends and the main intent here is to ensure that the values are meaningful. If we want to check that we have called the right UR APIs to implement those queries - that would be a `sycl/unittest` for sure and we can add that if necessary alongside with the one proposed here.

File: sycl/test-e2e/VirtualMem/memory_granularity.cpp (hunk 3 above)
Comment (DYNIO-INTEL) [reply sycl/test-e2e/VirtualMem/memory_granularity.cpp]: Done

**Comment by maarquitos14** [comment]
Thanks for the clarification, @AlexeySachkov, that makes sense.

**Review by maarquitos14 — APPROVED** [review]
LGTM, just a nit.

File: sycl/test-e2e/VirtualMem/memory_granularity.cpp [hunk 5]
```diff
@@ -0,0 +1,48 @@
...
+// granularity for virtual memory extension.
+
+// RUN: %{build} -o %t.out
+// RUN: %{run} %t.out
+
+#include <sycl/detail/core.hpp>
+
+#include <sycl/ext/oneapi/virtual_mem/virtual_mem.hpp>
+
+#include <assert.h>
```
Comment (maarquitos14) [inline sycl/test-e2e/VirtualMem/memory_granularity.cpp:12]: Most sycl-e2e tests go for `cassert` instead of `assert.h`. I think we should align.

Comment (DYNIO-INTEL) [reply sycl/test-e2e/VirtualMem/memory_granularity.cpp:12]: Sure

**Review by maarquitos14 — APPROVED** [review]

**Comment by DYNIO-INTEL** [comment]
@intel/llvm-gatekeepers I would appreciate if this PR could be merged.

**Pull request merged**